src/http/matcher.cc

```diff
@@ -65,7 +66,12 @@ size_t param_matcher::match(const sstring& url, size_t ind, parameters& param) {
         }
         return sstring::npos;
     }
-    param.set(_name, url.substr(ind, last - ind));
+    sstring path;
+    if (http::internal::url_decode(url.substr(ind, last - ind), path)) {
+        param.set(_name, path);
```

Comment:
This is a breaking change because servers using this code (eg. redpanda here: https://github.com/redpanda-data/redpanda/blob/29be93c2b379e5b72afeb07eeed9db93cb2978df/src/v/redpanda/admin/security.cc#L269) that are currently assuming that param[] returns a raw undecoded path parameter are doing their own path decoding of the result of param[]. Without the server being aware of this breaking change, this could lead to double-decoding of path parameters, which is a problem because path decoding is not idempotent.

For more context, see these two comments:
* https://github.com/scylladb/seastar/pull/2125#discussion_r1547811750
* https://github.com/scylladb/seastar/pull/2125#discussion_r1542606401

Reply:
The situation here is very depressing. CC @avikivity @amnonh:

Basically, we have a library, Seastar HTTPD, with a serious bug - unlike what the HTTP RFC says a server needs to do, it does URL decoding in one case but not in other cases. This bug was known for more than 4 years (https://github.com/scylladb/seastar/issues/725) and caused application bugs, and yet the two major (only?) users of the library decided to work around the bug instead of just fixing it:
1. Scylla avoided using characters like "#" that can expose this bug (https://github.com/scylladb/scylladb/issues/5883)
2. Redpanda worked around the bug by doing URL decoding on the result that the library forgot to do it.

Now that someone (me) got fed up with this bug and wanted to fix it, he is told: No, we need to keep the bug because users now rely on the buggy behavior. We can't fix the bug and will need to keep it forever.

I don't want to accept that. If the consensus is that we can't fix param[], then we need to deprecate it and introduce a new, bug-free, API. We shouldn't keep both as good alternatives - the buggy one isn't good - it's buggy.

Reply:
Specifically for the query params, I vote for deprecating and using an alternative correct implementation. When doing so, changing its API, the current implementation returns a single result per param (which is easy to use), but query parameters can support multiple values, if we are fixing things, we can fix this as well.

Reply:
The splitting of the "query" part of the different keys and values is defined by the HTML standard (not HTTP or URL standards), and needs to happen before the URL decoding (yes, it's a big mess...). I'm not aware of similar encoding for the "path" part. The path part is normally just a single string with slashes in it, and doesn't have key-value pairs. Or maybe you meant to separate it into a vector, according to the un-encoded slashes? That could be done, but I don't know if anyone will need it, and it will complicate the lives of current users who get the entire path part unsplit (unless I'm missing something).